### PR TITLE
Video upload (take 2)

### PR DESCRIPTION
**Pull request by jamesandres**
This is a more up to date version of https://github.com/tweepy/tweepy/pull/672 originally crafted by @Choko256 and @fitnr.

The original description from the previous PR:
> As I mentioned in #655, this is an alternate way of adding video upload. The media_upload method gets a conditional, with smaller images routed to 'normal' upload, and larger images and videos chunked upload. I also moved upload limits to a class variable, to make them easier to change if the need arises.

> This also fixes from unicode errors as flagged by @jeremylow.

### CODE REVIEW

**Comment by jamiewinspear** [comment]
I have found a bug with this - It currently only allows support for chunked media uploads related to public content. As it turns out the media_category parameter is not differentiated solely by content type ("image", "video" or "gif") but by whether the content is public or private also. It can also take values of dm_image, dm_gif or dm_video. I will be looking into fixing this tomorrow, but would hesitate to merge this before it is addressed.

**Comment by TheEssemCraft** [comment]
It's been addressed, now what?

**Comment by jamiewinspear** [comment]
At this point we are just waiting for a merge, we have been using this in production from the time I committed that last fix with no issues.

**Comment by TheEssemCraft** [comment]
Maybe you can merge the changes from the main repo into your fork? That
might give it a better chance of being merged.

**Comment by fitnr** [comment]
@TheEssemCraft There aren't any conflicts between this branch and `origin/master`. There have only been three pull requests merged since this branch was rebased onto `master`, all of them relatively minor.

**Comment by TheEssemCraft** [comment]
Alright, now all we have to do is wait...

**Comment by fitnr** [comment]
@jamesandres @jamiewinspear 
I've found and fixed a bug when uploading images from file handles. The name of the parameter was inconsistent (`file` or `f`). While fixing, I found it useful to move the functionality for checking filesize to a function in `utils.py`. The [update is in my fork](https://github.com/fitnr/tweepy/tree/video_upload2), I suggest that the commits should be added to this pull.

**Comment by guilouro** [comment]
I've been tried to use this PR, but when I try to do `stream` I receive this error:

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.6/site-packages/tweepy/streaming.py", line 297, in _run
    six.reraise(*exc_info)
  File "/usr/local/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.6/site-packages/tweepy/streaming.py", line 266, in _run
    self._read_loop(resp)
  File "/usr/local/lib/python3.6/site-packages/tweepy/streaming.py", line 323, in _read_loop
    raise TweepError('Expecting length, unexpected value found')
tweepy.error.TweepError: Expecting length, unexpected value found
```
Is it happened with somebody else?

**Comment by fitnr** [comment]
The only difference between the two branches that affects streaming is 176061133bd, a commit added to `tweepy:master` since this pull request was made. Try creating a branch with both sets of changes:
```
git checkout -b video_merge video_upload2
git merge tweepy/master
python setup.py install
```

**Comment by guilouro** [comment]
Thanks for you answer @fitnr ... I solved using `streaming.py` of the master branch.

**Comment by braian87b** [comment]
Please merge this thing...

**Comment by braian87b** [comment]
https://github.com/tweepy/tweepy/pull/672 is from year 2015... 🤦‍♂️

**Comment by Xpktro** [comment]
I just stumbled upon this when implementing an automated video upload. I would be really grateful if you merge this so I won't stop using this library!

**Comment by braian87b** [comment]
@Aaron1011, @joshthecoder Hi guys, sorry to bother, but, is there an issue with this?

**Comment by braian87b** [comment]
@Xpktro if you want use my fork at https://github.com/braian87b/tweepy, it has the video upload feature of this PR merged, to use it:
```
upload_result = api.media_upload('/home/user/video.mp4')
api.update_status(status="test tweet", media_ids=[upload_result.media_id_string])
```

**Comment by corbindavenport** [comment]
@braian87b Is there any way you could upload your forked version to the [Python Package Index](https://pypi.python.org/pypi)?

**Comment by braian87b** [comment]
Yes there is, but I will not do it yet, lets wait a few more weeks, main devs could be on vacation or something, I dont think that this is on abandanware status yet to do that (but I may be wrong). If you want to do it feel free, the main code, and the PR code are not mine, I just make the repo to be easily to download the files and not being patching them at hand.

**Comment by jh69** [comment]
i've been using braian87b's video upload code and it seems to work pretty well. any idea when we can get this put up?

**Comment by braian87b** [comment]
Twitter.com will start complaining about the avalanche videos of cats/memes/etc from apps and bots.

**Comment by gloriouskilka** [comment]
bump

**Comment by gloriouskilka** [comment]
It this ok to not to upload videos and having this PR in 2018 for popular twitter python package?

**Comment by fitnr** [comment]
@joshthecoder Any thoughts on merging this pull?

**Comment by braian87b** [comment]
@RedHazardTwelve  @fitnr  just use my fork, it already have the code needed to upload videos: https://github.com/tweepy/tweepy/pull/929#issuecomment-362914398

**Comment by fitnr** [comment]
@RedHazardTwelve I co-wrote the code, I'm asking on behalf of others.

**Comment by corbindavenport** [comment]
@braian87b but then you have to keep maintaining that fork, and it makes installation more difficult because it's not available from PIP. The dev of this library just merged a PR nine days ago, I'm confused why this one hasn't been accepted.

@joshthecoder is there anything else that needs to be done before this can be merged?

**Comment by RedHazardTwelve** [comment]
@braian87b right, but how do I get the media_id_string, I can't seem to find it in the documentation? Also where does it go (before or after reply_id)?

**Comment by braian87b** [comment]
@corbindavenport no, I'm not 😆 
Make your own fork and request a package on pypi (pip) for them to have a `pip install tweepy-fork` available.

**Comment by Spyder-exe** [comment]
Right so I used the fork of @braian87b And it seems to work rather well 
I even changed a few things (actually only one line) so that it's compatible with the latest pip version
(btw @corbindavenport if even it doesn't change the maintenance issue, you can just use pip by doing
``pip install git+//github.com/Spyder-exe/tweepy.git
And with the original tweepy uninstalled, it works perfectly fine.

However I seem to have two issues:
**Issue number one:** Error 324 ('Not valid video') which after some digging around means that the video is or too short or too long, however my video is only 26 seconds long and 650kb when the API should allow up to 15 MB max?
**Issue number 2:** 'Failed to send request:[...], Max retries exceeded with url: ...' which within of itself is self explanatory however I was only trying to send 1 video every 3 minutes which should be within the API limitations?

**Comment by iandow** [comment]
@Spyder-exe Thanks for the fork. One fix: binder.py needs to allow 202 and 204 status codes in order to avoid error "TypeError: argument of type 'int' is not iterable". I submitted a pull request against your fork.

**Comment by mjarrett** [comment]
@Spyder-exe Thanks for making that fork... curious if you ever figured out what's going with your issue number one? The videos I'm making upload fine through the web interface but I get the 324 error with the API.

**Comment by alsuren** [comment]
I just rebased on top of master and force-pushed. The only conflict was in a docstring.

We have been using this in prod since 2017 and it seems fine. I will try to deploy this rebased version tomorrow.

We have a little wrapper around it that shouldn't be contributed upstream, but is worth mentioning here:

```

def _media_upload(connection, name, file_obj=None, is_direct_message=False):
    """Wrap API connection.media_upload to check for readiness

    Media that are to be attached to Tweets must first be uploaded to Twitter.
    If one attempts to use the media before Twitter has finished processing
    it then the result is an internal server error. The proper solution would
    be to check whether or not a media upload has completed before using the
    media.

    When a media_upload call is made to the Twitter API we check for a
    'pending' or 'in_progress' state. If the request hasn't completed then we
    wait for the suggested time (called 'check_after_secs') before returning
    the media reference.

    What could still go wrong? The request could still be pending or
    in_progress after the delay, and this doesn't recheck the status. To be
    able to check the status we need to be able to call GET media/upload
    (STATUS). The API library doesn't currently support this, and adding it
    here seems a hack too far.
    """

    # HACK: This should really be handled in the API library layer
    media = connection.media_upload(
        name, file=file_obj, is_direct_message=is_direct_message)
    blocking_statuses = ['pending', 'in_progress']
    if (hasattr(media, 'processing_info') and
            media.processing_info['state'] in blocking_statuses):
        check_after = media.processing_info['check_after_secs']
        time.sleep(check_after)

    return media
```

**Comment by umar13893** [comment]
Hi, any solution for this? My status updating on twitter profile without video, for the image it works fine.

**Comment by ders99** [comment]
Any updates on the "Not valid video" error? Every video I try to upload seems to run into this.

**Comment by umar13893** [comment]
Any updates on this? I am still looking to upload an MP4 video using Tweepy.

**Comment by Shumaister** [comment]
Hello there,
 
i read a lot of issues here and, I try this
```
upload_result = api.media_upload('TestVideo.mp4', file='/myPath/TestVideo.mp4')
api.update_status(status="Texto del tuit", media_ids=[upload_result.media_id_string])
```
And I've got this message:

> raise TweepError('Invalid file type for image: %s' % file_type)
> tweepy.error.TweepError: Invalid file type for image: video/mp4

i don't know if I have an error or this is a problem with tweepy

**Comment by mattdonders** [comment]
In Discord, it was asked to put together a MCVE for this so that's here now. This grabs an MP4 from the NHL back-end API and tries to post it to Twitter.

It requires you to pass in consumer & access keys, but should be readily available - 
```bash
python3 nhl-highlight-ripper-poster.py --twitter \
    --consumerkey "XXXXXXXX" --consumersecret "XXXXXXXX" \
    --accesstoken "XXXXXXXX" --accesssecret "XXXXXXXX"
```

```python
import argparse
import requests
import shutil
import tweepy


parser = argparse.ArgumentParser()
parser.add_argument("--discord", help="log to console instead of file", action="store_true")
parser.add_argument("--disordwebhook", help="override team in configuration", action="store")
parser.add_argument("--twitter", help="log to console instead of file", action="store_true")
parser.add_argument("--consumerkey", help="override team in configuration", action="store")
parser.add_argument("--consumersecret", help="override team in configuration", action="store")
parser.add_argument("--accesstoken", help="override team in configuration", action="store")
parser.add_argument("--accesssecret", help="override team in configuration", action="store")
args = parser.parse_args()


def download_file(url):
    local_filename = url.split("/")[-1]
    with requests.get(url, stream=True) as r:
        with open(local_filename, "wb") as f:
            shutil.copyfileobj(r.raw, f)

    return local_filename


sample_video = download_file(
    "https://hlslive-wsczoominwestus.med.nhl.com/publish/80b062f2-4778-4224-8dbb-df201e7eb233.mp4"
)

payload = {"content": "NHL highlight video."}
video_file = open(sample_video, "rb")
files = {"file": video_file}

if args.discord:
    response = requests.post(args.discordwebhook, files=files, data=payload)

if args.twitter:
    consumer_key = args.consumerkey
    consumer_secret = args.consumersecret
    access_token = args.accesstoken
    access_secret = args.accesssecret

    auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
    auth.set_access_token(access_token, access_secret)
    api = tweepy.API(auth)

    media_ids = [api.media_upload(sample_video).media_id_string]
    status = api.update_status(status="NHL highlight video.", media_ids = media_ids)

if not response.ok:
    print(response.json())
```

This is the error returned from code - 
```
Traceback (most recent call last):
  File "nhl-highlight-ripper-poster.py", line 50, in <module>
    media_ids = [api.media_upload(sample_video).media_id_string]
  File "/Users/mattdonders/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tweepy/api.py", line 200, in media_upload
    headers, post_data = API._pack_image(filename, 4883, form_field='media', f=f)
  File "/Users/mattdonders/.pyenv/versions/3.6.5/lib/python3.6/site-packages/tweepy/api.py", line 1322, in _pack_image
    raise TweepError('Invalid file type for image: %s' % file_type)
tweepy.error.TweepError: Invalid file type for image: video/mp4
```

Here is the `ffmpeg` output of the file - 
```
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '80b062f2-4778-4224-8dbb-df201e7eb233.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    encoder         : Lavf58.3.100
  Duration: 00:00:28.65, start: 0.000000, bitrate: 1335 kb/s
    Stream #0:0(und): Video: h264 (Main) (avc1 / 0x31637661), yuv420p, 640x360 [SAR 1:1 DAR 16:9], 1261 kb/s, 59.96 fps, 59.96 tbr, 90k tbn, 119.88 tbc (default)
    Metadata:
      handler_name    : VideoHandler
    Stream #0:1(und): Audio: aac (LC) (mp4a / 0x6134706D), 44100 Hz, stereo, fltp, 64 kb/s (default)
    Metadata:
      handler_name    : SoundHandler
```

File: tweepy/api.py [hunk 1]
```diff
@@ -1383,3 +1481,114 @@ def _pack_image(filename, max_size, form_field="image", f=None):
...
+        else:
+            prefix = 'tweet'
+
+        if file_type in IMAGE_MIMETYPES:
+            if file_type == 'image/gif':
+                return prefix + '_gif'
+            else:
+                return prefix + '_image'
+        elif file_type == 'video/mp4':
+                    return prefix + '_video'
```
Comment (Harmon758) [inline tweepy/api.py:1594]:
```suggestion
            return prefix + '_video'
```

File: tweepy/binder.py [hunk 2]
```diff
@@ -182,12 +182,20 @@ def execute(self):
...
+                                                    timeout=self.api.timeout,
+                                                    auth=auth,
+                                                    proxies=self.api.proxy)
+                    except UnicodeEncodeError:
+                        resp = self.session.request(self.method,
+                                                    full_url,
+                                                    data=self.post_data.decode('utf-8'),
+                                                    timeout=self.api.timeout,
+                                                    auth=auth,
+                                                    proxies=self.api.proxy)
```
Comment (Harmon758) [inline tweepy/binder.py:198]: Why is this necessary?

File: tweepy/api.py [hunk 3]
```diff
@@ -212,6 +245,70 @@ def media_upload(self, filename, *args, **kwargs):
...
+                api=self,
+                path='/media/upload.json',
+                method='POST',
+                payload_type='media',
+                allowed_param=[],
+                require_auth=True,
+                upload_api=True
+            )(*args, **kwargs)
+        else:
+            return media_info
```
Comment (Harmon758) [inline tweepy/api.py:310]: When does this happen?

File: tweepy/api.py [hunk 4]
```diff
@@ -212,6 +245,70 @@ def media_upload(self, filename, *args, **kwargs):
...
+
+        # If a media ID has been generated, we can send the file
+        if media_info.media_id:
+            # default chunk size is 1MB, can be overridden with keyword argument.
+            # minimum chunk size is 16K, which keeps the maximum number of chunks under 999
+            chunk_size = kwargs.pop('chunk_size', 1024 * 1024)
+            chunk_size = max(chunk_size, 16 * 2014)
+
+            fsize = os.path.getsize(filename)
+            nloops = int(fsize / chunk_size) + (1 if fsize % chunk_size > 0 else 0)
```
Comment (Harmon758) [inline tweepy/api.py:281]:
```suggestion
            nloops = fsize // chunk_size + (fsize % chunk_size > 0)
```

File: tweepy/api.py [hunk 5]
```diff
@@ -1383,3 +1481,114 @@ def _pack_image(filename, max_size, form_field="image", f=None):
         }
 
         return headers, body
+
+    @staticmethod
+    def _chunk_media(command, filename, max_size, form_field="media", chunk_size=4096, f=None, media_id=None, segment_index=0, is_direct_message=False):
```
Comment (Harmon758) [inline tweepy/api.py:1486]: The default `chunk_size` should be consistent here.

File: tweepy/api.py [hunk 6]
```diff
@@ -1383,3 +1481,114 @@ def _pack_image(filename, max_size, form_field="image", f=None):
...
+        fp = None
+        if command == 'init':
+            if f is None:
+                file_size = os.path.getsize(filename)
+                try:
+                    if file_size > (max_size * 1024):
+                        raise TweepError('File is too big, must be less than %skb.' % max_size)
+                except os.error as e:
```
Comment (Harmon758) [inline tweepy/api.py:1494]:
```suggestion
                except OSError as e:
```

File: tweepy/api.py [hunk 7]
```diff
@@ -1383,3 +1481,114 @@ def _pack_image(filename, max_size, form_field="image", f=None):
...
+                    raise TweepError('Unable to access file: %s' % e.strerror)
```
Comment (Harmon758) [inline tweepy/api.py:1495]: `os.path.getsize` is outside the `try` block and nothing else in it will raise `OSError`.

File: tweepy/api.py [hunk 8]
```diff
@@ -1383,3 +1481,114 @@ def _pack_image(filename, max_size, form_field="image", f=None):
...
+                # build the mulitpart-formdata body
+                fp = open(filename, 'rb')
+            else:
+                f.seek(0, 2)  # Seek to end of file
+                file_size = f.tell()
+                if file_size > (max_size * 1024):
+                    raise TweepError('File is too big, must be less than %skb.' % max_size)
+                f.seek(0)  # Reset to beginning of file
+                fp = f
+        elif command != 'finalize':
```
Comment (Harmon758) [inline tweepy/api.py:1506]: The only other option is
```suggestion
        elif command == 'append':
```

File: tweepy/api.py [hunk 9]
```diff
@@ -1383,3 +1481,114 @@ def _pack_image(filename, max_size, form_field="image", f=None):
...
+                raise TweepError('File input for APPEND is mandatory.')
+
+        # video must be mp4
+        file_type, _ = mimetypes.guess_type(filename)
+
+        if file_type is None:
+            raise TweepError('Could not determine file type')
+
+        if file_type not in CHUNKED_MIMETYPES:
+            raise TweepError('Invalid file type for video: %s' % file_type)
```
Comment (Harmon758) [inline tweepy/api.py:1519]: This was already determined in `media_upload` and could be an image.
The file type should be passed to this method rather than redetermined here.

File: tweepy/api.py [hunk 10]
```diff
@@ -1383,3 +1481,114 @@ def _pack_image(filename, max_size, form_field="image", f=None):
...
+                raise TweepError('Media ID is required for FINALIZE command.')
+            body.append(
+                urlencode({
+                    'command': 'FINALIZE',
+                    'media_id': media_id
+                }).encode('utf-8')
+            )
+            headers = {
+                'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8'
+            }
```
Comment (Harmon758) [inline tweepy/api.py:1570]: This can be combined with the prior if-elif statement.

**Comment by Harmon758** [comment]
@jamesandres @jamiewinspear @alsuren I've gone ahead and done a first pass at a review of the PR.

> I've found and fixed a bug when uploading images from file handles. The name of the parameter was inconsistent (`file` or `f`). While fixing, I found it useful to move the functionality for checking filesize to a function in `utils.py`. The [update is in my fork](https://github.com/fitnr/tweepy/tree/video_upload2), I suggest that the commits should be added to this pull.

There also seem to be further improvements by @fitnr that haven't been included in this PR yet.

@braian87b @jh69 @corbindavenport @Spyder-exe @iandow @mjarrett You can simply install from this PR or the branch of the fork this PR is from rather than creating or using another fork with the same changes.

@umar13893 @awcarlsson Do your videos meet the criteria in the Video specifications and recommendations section of https://developer.twitter.com/en/docs/media/upload-media/uploading-media/media-best-practices?

@Shumaister @mattdonders Did you install Tweepy from this PR?
@mattdonders Line 1322 of api.py where your error occurs isn't even part of `_pack_image` in this PR, so you're likely using a version that's not this PR.

**Comment by jackyliang** [comment]
Is this safe to pull to use prior to the v4 launch?

**Comment by Harmon758** [comment]
There shouldn't be any risk to using this PR, but there are still issues that need to be resolved.
Feel free to test it and report any issues you encounter.
Note though, that this PR seems to be stale now, and [I'm considering working off of it to create a new branch/PR](https://github.com/tweepy/tweepy/issues/640#issuecomment-671245434).

See https://tweepy.readthedocs.io/en/latest/install.html and https://pip.pypa.io/en/stable/reference/pip_install/#git.

File: tweepy/api.py [hunk 11]
```diff
@@ -195,11 +205,34 @@ def update_status(self, *args, **kwargs):
         )(post_data=post_data, *args, **kwargs)
 
     def media_upload(self, filename, *args, **kwargs):
-        """ :reference: https://developer.twitter.com/en/docs/media/upload-media/api-reference/post-media-upload
+        """ :reference: https://dev.twitter.com/rest/reference/post/media/upload
+            :allowed_param:
+        """
+        f = kwargs.pop('file', None)
+
+        mime, _ = mimetypes.guess_type(filename)
```
[the remark this answers is not in the document]
Comment (fitnr) [reply tweepy/api.py:213]: `imghdr` doesn't work on video files, so that's another reason to keep it as a backup.

**Comment by Harmon758** [comment]
Superseded by #1414

**Comment by Harmon758** [comment]
The video-upload branch / pull request #1486 should be complete now.
Any feedback or review would be appreciated.
